File: examples/dance.hs
-- At its core this algorithm is working in an incredibly similar way to the
-- Rust version. The biggest difference is the actual order the operations come
-- in the syntax. Haskell is a little different from most other languages in
-- that it doesn't have methods (functions that are defined on a type), rather
-- it exclusively relies on functions. Because of this the order the operation
-- that is ran first should be furthest right, and the one that runs first
-- should be furthest left. Another weird askect is that functions aren't
-- called using parenthesis, rather they are called by passing arguments to
-- them with spaces to seperate them. For example (+) 1 1 would call the +
-- function on the parameters 1 and 1, returning 2. Of course there are also 

mrRoboto :: [String] -> (Int, Int)
mrRoboto instructions =
    foldl (\(acc_l, acc_r) (cur_l, cur_r) -> (acc_l + cur_l, acc_r + cur_r)) (0, 0)
    $ map (dirToIntPair . words) instructions
  where dirToIntPair [dir, strVal] =
          let val = read strVal :: Int
          in case dir of
            "up"    -> (0, val)
            "down"  -> (0, -val)
            "right" -> (val, 0)
            "left"  -> (-val, 0)

main :: IO()
main = do
    let t1 = mrRoboto ["right 10", "up 50", "left 30", "down 10"]
    putStrLn $ "test1: " ++ show t1
    let t2 = mrRoboto ["right 100", "right 100", "up 500", "up 10000"]
    putStrLn $ "test2: " ++ show t2
    let t3 = mrRoboto []
    putStrLn $ "test3: " ++ show t3

File: examples/dance.rs
// A robot has been given a list of movement instructions. Each instruction
// is either left, right, up or down, followed by a distance to move. The
// robot starts at 0, 0. You want to calculate where the robot will end up
// and return its final position as a list. For example, if the robot is given
// the instructions ["right 10", "up 50", "left 30", "down 10"], it will end
// up 20 left and 40 up from where it started, so you should return [-20, 40].

fn mr_roboto(instructions: &[&str]) -> (i32, i32) {
    // At its core this implementation is doing almost exactly the same actual processing of the
    // input as the second Python functional implementation. The biggest difference is that the
    // steps can be broken up in the implementation into more easily understandable steps. This is
    // because Rust has a thoughtfully implementated Lazy iterators.
    // With that being said, what do you think is the Big O time complexity of this
    // algorithm? In other words, how many times do we need to iterate over a set of data to
    // generate the needed results?
    return instructions
        .iter()
        .map(|d_val_str| -> (&str, i32) {
            let d_val_arr: Vec<&str> = d_val_str.split_whitespace().collect();
            println!("{} {}", d_val_arr[0], d_val_arr[1]);
            // Here we are costructing a tuple out of the first element and the
            // second element, which is also converted into a 32bit int.
            return (d_val_arr[0], d_val_arr[1].parse().unwrap());
        })
        .map(|(direction, val)| {
            println!("{} {}", direction, val);
            match direction {
                "up" => (0, val),
                "down" => (0, -val),
                "right" => (val, 0),
                "left" => (-val, 0),
                _ => panic!(),
            }
        })
        .fold((0, 0), |(acc_l, acc_r), (cur_l, cur_r)| {
            (acc_l + cur_l, acc_r + cur_r)
        });
}

fn main() {
    let test_1 = mr_roboto(&["right 10", "up 50", "left 30", "down 10"]);
    println!("test1: {:?}", test_1);
    let test_2 = mr_roboto(&["right 100", "right 100", "up 500", "up 10000"]);
    println!("test2: {:?}", test_2);
    let test_3 = mr_roboto(&[]);
    println!("test3: {:?}", test_3);
}
